### PR TITLE
fix SourceFile.lineContent crashes for position at end of line

### LINKED ISSUE #785: SourceFile.lineContent crashes for position at end of line
SourceFile:lineContent crashes in tokenEndCodePointIndex when the position is position following last character in file.

For example, if the line contains just a close brace and the position is 1. In this case, it should return `["}", "", ""]`.

(We probably shouldn't ever do this, but still better not to crash here.)

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose

- Fix SourceFile.lineContent crashes for position at end of line
- Provide meaningful error messages for out of range positions

Resolves #785

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/token.bal [hunk 1]
```diff
@@ -569,6 +571,19 @@ public readonly class SourceFile {
     function scannedLine(int lineNumber) returns ScannedLine {
         return self.lines[lineNumber - 1];
     }
+
+    function validatePosition(Position pos) {
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: This is not the right way to solve this problem.  Please do something more like what I suggested in the issue.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/token.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/token.bal]: fixed

**Comment by jclark** [comment]
There seems to be a misunderstanding about what the issue is.

The issue is #785, which is "SourceFile.lineContent crashes for position at end of line".  The issue  is not "Provide meaningful error messages for out of range positions".

Fixing #785 means fixing lineContent so that e.g. if the line contains just a close brace and the position is 1, the lineContent should return ["}", "", ""].

The 2nd part of this is that we should not be constructing errors wiith locations where lineContent would return a 2nd string that is empty. So when d.format gets such a return value from lineContent, it should do so in a way that ensures we know that there's a problem and can fix it (that is related to the subject of this PR, but is a specific case).

**Review by jclark — CHANGES_REQUESTED** [review]
Please also fix the PR title.

**Comment by heshanpadmasiri** [comment]
> There seems to be a misunderstanding about what the issue is.
> 
> The issue is #785, which is "SourceFile.lineContent crashes for position at end of line". The issue is not "Provide meaningful error messages for out of range positions".
> 
> Fixing #785 means fixing lineContent so that e.g. if the line contains just a close brace and the position is 1, the lineContent should return ["}", "", ""].
> 
> The 2nd part of this is that we should not be constructing errors wiith locations where lineContent would return a 2nd string that is empty. So when d.format gets such a return value from lineContent, it should do so in a way that ensures we know that there's a problem and can fix it (that is related to the subject of this PR, but is a specific case).

fixed. (`lineContent` will no longer fail in such cases and d:format will panic with a meaningful error message in such cases (in addition to out of range positions)

File: compiler/modules/comm.diagnostic/diagnostic.bal [hunk 2]
```diff
@@ -28,6 +28,8 @@ public type File readonly & object {
     public function lineColumn(Position pos) returns LineColumn;
     public function lineContent(Position|Range range) returns [string, string, string];
     public function qNameRange(Position startPos) returns Range;
+    public function maxLineNum() returns int;
```
Comment (jclark) [inline compiler/modules/comm.diagnostic/diagnostic.bal]: This is not part of the issue.

File: compiler/modules/comm.diagnostic/diagnostic.bal [hunk 3]
```diff
@@ -108,6 +110,7 @@ public function messageFormat(Template t) returns string {
 
 public function format(Diagnostic d) returns string[] {
     Location loc = d.location;
+    validateLocation(loc);
```
Comment (jclark) [inline compiler/modules/comm.diagnostic/diagnostic.bal]: This is not part of the issue.

File: compiler/modules/comm.diagnostic/diagnostic.bal [hunk 4]
```diff
@@ -116,11 +119,37 @@ public function format(Diagnostic d) returns string[] {
...
+    string[] content = loc.file.lineContent(range);
+    if content[1] == "" {
+        panic error(string `error range ${lc.toString()} is empty in ${loc.file.filename()}`);
+    }
+    lines.push("".'join(...content));
     lines.push(caretLine(loc.file, range));
     return lines;
 }
 
+function validateLocation(Location loc) {
```
Comment (jclark) [inline compiler/modules/comm.diagnostic/diagnostic.bal]: This is not part of the issue.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/comm.diagnostic/diagnostic.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/comm.diagnostic/diagnostic.bal]: fixed

File: compiler/modules/comm.diagnostic/diagnostic.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/comm.diagnostic/diagnostic.bal]: fixed

File: compiler/modules/comm.diagnostic/diagnostic.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/comm.diagnostic/diagnostic.bal]: fixed

File: compiler/modules/front.syntax/token.bal [hunk 5]
```diff
@@ -547,6 +547,9 @@ public readonly class SourceFile {
                 endColumnNum = lineContent.length();
             }
         }
+        else if startColumnNum == lineContent.length() {
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: I would have thought it would be cleaner to fix tokenEndCodePointIndex...

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/token.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/token.bal]: fixed

**Pull request merged**